streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java

```diff
@@ -115,15 +117,22 @@ public void init(final ProcessorContext context,
         }
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public void close() {
-        super.close();
-        metrics.removeAllStoreLevelSensors(taskName, name());
-    }
-
-    @Override
-    public long approximateNumEntries() {
-        return wrapped().approximateNumEntries();
+    public boolean setFlushListener(final CacheFlushListener<K, V> listener,
+                                    final boolean sendOldValues) {
+        final KeyValueStore<Bytes, byte[]> wrapped = wrapped();
+        if (wrapped instanceof CachedStateStore) {
```

Comment:
We know the wrapping order, thus, this is safe. Intercepting the callback below to deserialize.

Reply:
Meta: I like this approach slightly more than the previous one, and wondering if we can sort-of combine these two to make the logic more readable (and also the metered store lost the topic name info when calling serde which may not be acceptable):

1) Metered store do not need to inherited CachedStateStore, instead inherit a SerdeStateStore interface which just expose the getter of the serdes (note we cannot use `serdes` since it is only constructed at initialization, we have to use `keySerde` and `valueSerde` instead`).
2) In TupleForwarder we get both the SerdeStateStore, and check if its wrapped store is a CachedStateStore, if yes call `cachedStore.setFlushListener(... listener.apply(serdeStore.keySerde().deser ...))` -- i.e. basically the code below.

WDYT?

Reply:
> and also the metered store lost the topic name info when calling serde which may not be acceptable

Fixed this part. Compare my other comments.

About your proposal: that is certainly possible, it's similar (or even the same?) to the original PR introducing `TypedStore` (your suggested `SerdeStateStore` is the same thing). So I am wondering what the difference to the original proposal would be -- seems you suggest to go back to the original one.

Reply:
I also think that using two nested `CacheFlushListener` is better instead of "drilling" whole into the APIs. So I prefer to stay with the current design.